File: haskell/3-signature.hs
fac :: Int -> Int
fac 0 = 1
fac 1 = 1
fac n = fac (n - 1) + fac (n - 2)

sorted :: [Int] -> Bool 
sorted [] = True 
sorted [_] = True
sorted (a:ys@(b:_)) = a < b && sorted ys

File: prolog/3-signature.pl
fac(0, 1) :- !.
fac(1, 1) :- !.
fac(X, Y) :-
    N1 is X - 1,
    N2 is X - 2,
    fac(N1, Y1),
    fac(N2, Y2),
    Y is Y1 + Y2.

sorted([]).
sorted([_]) :- !.
sorted([A,B|Xs]) :- A < B, sorted([B|Xs]).
